### PR TITLE
✨Add Context Menu to SolutionExplorer

### PR DESCRIPTION
**Pull request by Vyperus**
This PR adds a simple context menu to the Solution Explorer. In this stage the the menu only can copy existing diagrams in FS Solutions.

**Changes:**

1. Add a Basic Context Menu to the Solution Explorer.
2. Add the Copy Functionality (Right Click Copy) for FileSystem Solutions.

PR: #1416

## How can others test the changes?

1. Open a FS Solution.
2. Right click on a diagram.
3. Click on "Copy".

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

### CODE REVIEW

File: src/services/diagram-creation-service/DiagramCreationService.ts [hunk 1]
```diff
@@ -19,6 +26,42 @@ export class DiagramCreationService implements IDiagramCreationService {
     return diagram;
   }
 
+  private async _renameDiagram(xml: string, name: string): Promise<string> {
+    const modeler: IBpmnModeler = new bundle.modeler({});
+
+    modeler.importXML(xml, (error: any) => {
+      // Do nothing for now
```
Comment (ElRaptorus) [inline src/services/diagram-creation-service/DiagramCreationService.ts]: Not even a notification, if the import failed?

Comment (ElRaptorus) [inline src/services/diagram-creation-service/DiagramCreationService.ts]: Not even a notification, if the import failed?

I think the end user might want to know about a failed import. Or the reason for it.

File: src/services/diagram-creation-service/DiagramCreationService.ts [hunk 2]
```diff
@@ -19,6 +26,42 @@ export class DiagramCreationService implements IDiagramCreationService {
...
+    });
+
+    const promise: Promise<string> = new Promise((resolve: Function): void => {
```
Comment (ElRaptorus) [inline src/services/diagram-creation-service/DiagramCreationService.ts]:
```suggestion
    const promise: Promise<string> = new Promise((resolve: Function, reject: Function): void => {
```

See below as to why.

File: src/services/diagram-creation-service/DiagramCreationService.ts [hunk 3]
```diff
@@ -19,6 +26,42 @@ export class DiagramCreationService implements IDiagramCreationService {
...
+        const collaboration: IModdleElement = rootElements.find((element: IModdleElement) => {
+          return element.$type === 'bpmn:Collaboration';
+        });
+        const participant: IModdleElement = collaboration.participants[0];
+
+        participant.name = name;
+        participant.processRef = process;
+
+        modeler.saveXML({}, (error: Error, result: string) => {
+          resolve(result);
```
Comment (ElRaptorus) [inline src/services/diagram-creation-service/DiagramCreationService.ts]:
```suggestion
          const errorOccured: boolean = error !== undefined;
          if (errorOccured) {
            return reject(error);
          }
          resolve(result);
```

File: test/e2e/src/pages/diagramDetail.ts [hunk 4]
```diff
@@ -20,6 +20,8 @@ export class DiagramDetail {
   }
 
   public async clickOnElement(elementId: string): Promise<void> {
+    browser.wait(ExpectedConditions.visibilityOf(this._bpmnIoContainer), browser.params.defaultTimeoutMS);
```
Comment (ElRaptorus) [inline test/e2e/src/pages/diagramDetail.ts]: Based on what I've seen in other PRs, shouldn't this be:

```suggestion
    await browser.wait(ExpectedConditions.visibilityOf(this._bpmnIoContainer), browser.params.defaultTimeoutMS);
```

?

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.scss [hunk 5]
```diff
@@ -121,3 +121,21 @@
...
+  position: absolute;
+  background: white;
+  border: 1px solid #dcdbdb;
+  border-radius: 5px;
+}
+
+.diagram-contextmenu__entry {
+  width: 100%;
+  text-align: center;
+  padding: 0 10px;
```
Comment (alexanderkasten) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.scss]: Please move the padding between width and text-align

File: src/services/diagram-creation-service/DiagramCreationService.ts [hunk 6]
```diff
@@ -1,13 +1,30 @@
-import {IDiagram, ISolution} from '@process-engine/solutionexplorer.contracts';
-import {IDiagramCreationService} from '../../contracts';
+import {inject} from 'aurelia-framework';
 
+import {IModdleElement, IProcessRef} from '@process-engine/bpmn-elements_contracts';
+import * as bundle from '@process-engine/bpmn-js-custom-bundle';
+import {IDiagram} from '@process-engine/solutionexplorer.contracts';
+
+import {IBpmnModeler, IDiagramCreationService, NotificationType} from '../../contracts/index';
+import { NotificationService } from '../notification-service/notification.service';
```
Comment (alexanderkasten) [inline src/services/diagram-creation-service/DiagramCreationService.ts]: Remove whitespaces

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.scss [hunk 7]
```diff
@@ -121,3 +121,21 @@
...
 }
+
+.diagram-contextmenu {
+  position: absolute;
+  background: white;
+  border: 1px solid #dcdbdb;
+  border-radius: 5px;
+}
+
+.diagram-contextmenu__entry {
```
Comment (SteffenKn) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.scss]: Please sort this according to the css guideline

```
.diagram-contextmenu__entry {
  width: 100%;
  padding: 0 10px;
  text-align: center;
}
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 8]
```diff
@@ -250,6 +253,48 @@ export class SolutionExplorerSolution {
...
+    let diagramNumber: number = 1;
+
+    while (newNameFound === false) {
+      newName = `${this._diagramInContextMenu.name} (${diagramNumber})`;
+
+      newNameFound = this.openedDiagrams.every((diagram: IDiagram) => {
+        return diagram.name !== newName;
+      });
+
+      diagramNumber += 1;
```
Comment (SteffenKn) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]: IMO this is easier to read:

```suggestion
      diagramNumber++;
```

File: src/services/diagram-creation-service/DiagramCreationService.ts [hunk 9]
```diff
@@ -19,6 +36,48 @@ export class DiagramCreationService implements IDiagramCreationService {
...
+    const modeler: IBpmnModeler = new bundle.modeler({});
+
+    modeler.importXML(xml, (error: Error) => {
+      this._notificationService.showNotification(NotificationType.ERROR, `Failed to copy diagram. ${error.message}`);
+    });
+
+    const promise: Promise<string> = new Promise((resolve: Function, reject: Function): void => {
+      modeler.on('import.done', () => {
+
+        const rootElements: Array<any> = modeler._definitions.rootElements;
```
Comment (alexanderkasten) [inline src/services/diagram-creation-service/DiagramCreationService.ts]:
```suggestion
        const rootElements: Array<IModdleElement>  = modeler._definitions.rootElements;
```

File: src/services/diagram-creation-service/DiagramCreationService.ts [hunk 10]
```diff
@@ -19,6 +36,48 @@ export class DiagramCreationService implements IDiagramCreationService {
...
+        const process: IProcessRef = rootElements.find((element: any) => {
```
Comment (alexanderkasten) [inline src/services/diagram-creation-service/DiagramCreationService.ts]:
```suggestion
        const process: IProcessRef = rootElements.find((element: IModdleElement) => {
```

File: src/services/diagram-creation-service/DiagramCreationService.ts (hunk 6 above)
Comment (SteffenKn) [inline src/services/diagram-creation-service/DiagramCreationService.ts]:
```suggestion
import {NotificationService} from '../notification-service/notification.service';
```

File: src/services/diagram-creation-service/DiagramCreationService.ts [hunk 11]
```diff
@@ -19,6 +36,48 @@ export class DiagramCreationService implements IDiagramCreationService {
...
+        });
+        const participant: IModdleElement = collaboration.participants[0];
+
+        participant.name = name;
+        participant.processRef = process;
+
+        modeler.saveXML({}, (error: Error, result: string) => {
+          const errorOccured: boolean = error !== undefined;
+          if (errorOccured) {
+            this._notificationService.showNotification(NotificationType.ERROR, `Failed to copy diagram. ${error.message}`);
```
Comment (SteffenKn) [inline src/services/diagram-creation-service/DiagramCreationService.ts]: Add some spacing

```suggestion
            this._notificationService.showNotification(NotificationType.ERROR, `Failed to copy diagram. ${error.message}`);

```

**Comment by SteffenKn** [comment]
> Screenshot 2019-03-13 at 09 27 24
> 
> The context menu is in the wrong position if there is a scrollbar.

@Vyperus if you change the position in
```
.diagram-contextmenu {
  position: absolute;
  background: white;
  border: 1px solid #dcdbdb;
  border-radius: 5px;
}
```
to fixed, this will be fixed.

But you have to adjust the offset.

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.scss [hunk 12]
```diff
@@ -121,3 +121,21 @@
   width: calc(100% + 30px);
   margin-left: -30px;
 }
+
+.diagram-contextmenu {
+  position: absolute;
```
Comment (SteffenKn) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.scss]:
```suggestion
  position: fixed;
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 13]
```diff
@@ -250,6 +253,47 @@ export class SolutionExplorerSolution {
...
+      this._diagramInContextMenu = undefined;
+
+      document.removeEventListener('click', documentEventListener);
+    };
+
+    document.addEventListener('click', documentEventListener);
+  }
+
+  public async copyDiagram(): Promise<void> {
+    if (!this._diagramInContextMenu) { return; }
```
Comment (robpal94) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
    if (!this._diagramInContextMenu) {
    
    return;
  }
```

Also, please extract the evaluation rule here.

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 14]
```diff
@@ -250,6 +253,47 @@ export class SolutionExplorerSolution {
...
+    let diagramNumber: number = 1;
+
+    while (newNameFound === false) {
+      newName = `${this._diagramInContextMenu.name} (${diagramNumber})`;
+
+      newNameFound = this.openedDiagrams.every((diagram: IDiagram) => {
+        return diagram.name !== newName;
+      });
+
+      diagramNumber++;
```
Comment (robpal94) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
      diagramNumber += 1;
```

File: src/services/diagram-creation-service/DiagramCreationService.ts [hunk 15]
```diff
@@ -19,6 +36,49 @@ export class DiagramCreationService implements IDiagramCreationService {
     return diagram;
   }
 
+  private async _renameDiagram(xml: string, name: string): Promise<string> {
+    const modeler: IBpmnModeler = new bundle.modeler({});
+
+    modeler.importXML(xml, (error: Error) => {
+      this._notificationService.showNotification(NotificationType.ERROR, `Failed to copy diagram. ${error.message}`);
```
Comment (robpal94) [inline src/services/diagram-creation-service/DiagramCreationService.ts]:
```suggestion
      this._notificationService.showNotification(NotificationType.ERROR, `Failed to copy the diagram. Cause:  ${error.message}`);
```

**Comment by robpal94** [comment]
I would spawn the menu box a bit more to the right. This way, it wont intersect with the mouse cursor and the text is a bit better readable.

**Comment by Vyperus** [comment]
@robpal94 I think the menu is just right there. If I take a look at other context menus the menu is somewhat further left than right.

**Pull request merged**